icommand.lisp

```diff
@@ -0,0 +1,61 @@
+;; Copyright (C) 2016 Caio Oliveira
+;;
+;;  This file is part of stumpwm.
+;;
+;; stumpwm is free software; you can redistribute it and/or modify
+;; it under the terms of the GNU General Public License as published by
+;; the Free Software Foundation; either version 2, or (at your option)
+;; any later version.
+
+;; stumpwm is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with this software; see the file COPYING.  If not, see
+;; <http://www.gnu.org/licenses/>.
+
+;; Commentary:
+;;
+;;; Macro for defining interactive command. Just pushes and pops new keymaps.
+;;
+;; Code:
+
+(in-package #:stumpwm)
+
+(export '(icommand-enter-interactive-mode
+          icommand-exit-interactive-mode
+          deficommand))
+
+(defun icommand-enter-interactive-mode (kmap name)
+  "Enters interactive mode" ;; TODO
+  (message (format nil "~s started" name))
+  (push-top-map kmap))
+
+(defun icommand-exit-interactive-mode (name)
+  "Exits imove-focus mode"
+  (message "~s finished" name)
+  (pop-top-map))
+
+(defmacro deficommand (name key-bindings &key pre-start-fn cleanup-fn)
+  "Declare an interactive command." ;; TODO
+  (let* ((command (if (listp name) (car name) name))
+         (exit-command (concatenate 'string "exit-" (symbol-name command)))
+         (m-name (gensym "m")))
+    `(let ((,m-name (make-sparse-keymap)))
+       ,@(loop for keyb in key-bindings
+            collect `(define-key ,m-name ,@keyb))
+       (define-key ,m-name (kbd "RET") ,exit-command)
+       (define-key ,m-name (kbd "C-g") ,exit-command)
+       (define-key ,m-name (kbd "ESC") ,exit-command)
+       (defcommand ,name () ()
+         ,(concatenate 'string "Starts interactive command \"" ,(symbol-name command) "\"") ;; TODO Add docstring support
```

Comment:
There is an extra comma before symbol-name. The code fails to compile due to that.